File: src/web/Utils/CustomConverters.cs
using System;
using System.Text.Json;
using System.Text.Json.Serialization;
using core.fs;
using core.fs.Accounts;
using core.fs.Adapters.Brokerage;
using core.fs.Adapters.Stocks;
using core.fs.Alerts;
using core.fs.Options;
using core.fs.Reports;
using core.fs.Services;
using core.fs.Services.Analysis;
using core.fs.Stocks;
using core.Shared;
using core.Stocks;

namespace web.Utils;

public abstract class GenericConverterWithToString<T> : JsonConverter<T>
{
    public override void Write(Utf8JsonWriter writer, T value, JsonSerializerOptions options)
    {
        writer.WriteStringValue(value.ToString());
    }
}

public class TickerConverter : JsonConverter<Ticker>
{
    public override Ticker Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        return new Ticker(reader.GetString());
    }

    public override void Write(Utf8JsonWriter writer, Ticker value, JsonSerializerOptions options)
    {
        writer.WriteStringValue(value.Value);
    }
}

public class OptionTickerConverter : JsonConverter<OptionTicker>
{
    public override OptionTicker Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        return OptionTicker.NewOptionTicker(reader.GetString());
    }

    public override void Write(Utf8JsonWriter writer, OptionTicker value, JsonSerializerOptions options)
    {
        writer.WriteStringValue(value.Item);
    }
}

public class UserIdConverter : JsonConverter<UserId>
{
    public override UserId Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        return UserId.NewUserId(new Guid(reader.GetString()));
    }

    public override void Write(Utf8JsonWriter writer, UserId value, JsonSerializerOptions options)
    {
        writer.WriteStringValue(value.Item);
    }
}

public class GapTypeConverter : GenericConverterWithToString<GapAnalysis.GapType>
{
    public override GapAnalysis.GapType Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        return GapAnalysis.GapType.FromString(reader.GetString());
    }
}

public class PriceFrequencyConverter : GenericConverterWithToString<PriceFrequency>
{
    public override PriceFrequency Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        return PriceFrequency.FromString(reader.GetString());
    }
}

public class OutcomeReprtDurationConverter : GenericConverterWithToString<OutcomesReportDuration>
{
    public override OutcomesReportDuration Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        return OutcomesReportDuration.FromString(reader.GetString());
    }
}

public class OptionTypeConverter : GenericConverterWithToString<OptionType>
{
    public override OptionType Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        return OptionType.FromString(reader.GetString());
    }
}

public class GradeConverter : JsonConverter<TradeGrade>
{
    public override TradeGrade Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        return new TradeGrade(reader.GetString());
    }

    public override void Write(Utf8JsonWriter writer, TradeGrade value, JsonSerializerOptions options)
    {
        writer.WriteStringValue(value.Value);
    }
}

public class ValueFormatTypeConverter : GenericConverterWithToString<ValueFormat>
{
    public override ValueFormat Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        return ValueFormat.FromString(reader.GetString());
    }
}

public class OutcomeTypeTypeConverter : GenericConverterWithToString<OutcomeType>
{
    public override OutcomeType Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        return OutcomeType.FromString(reader.GetString());
    }
}

public class BrokerageOrderTypeConverter : GenericConverterWithToString<BrokerageOrderType>
{
    public override BrokerageOrderType Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        return BrokerageOrderType.FromString(reader.GetString());
    }
}

public class BrokerageOrderDurationConverter : GenericConverterWithToString<BrokerageOrderDuration>
{
    public override BrokerageOrderDuration Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        return BrokerageOrderDuration.FromString(reader.GetString());
    }
}

public class PositionEventTypeConverter : JsonConverter<PositionEventType>
{
    public override PositionEventType Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        return new PositionEventType(reader.GetString());
    }

    public override void Write(Utf8JsonWriter writer, PositionEventType value, JsonSerializerOptions options)
    {
        writer.WriteStringValue(value.Value);
    }
}

public class ChartAnnotationLineTypeConverter : GenericConverterWithToString<ChartAnnotationLineType>
{
    public override ChartAnnotationLineType Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        return ChartAnnotationLineType.FromString(reader.GetString());
    }
}

public class DataPointChartTypeConverter : GenericConverterWithToString<DataPointChartType>
{
    public override DataPointChartType Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        return DataPointChartType.FromString(reader.GetString());
    }
}

public class StockPositionIdConverter : JsonConverter<StockPositionId>
{
    public override StockPositionId Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        return StockPositionId.NewStockPositionId(Guid.Parse(reader.GetString()));
    }
    
    public override void Write(Utf8JsonWriter writer, StockPositionId value, JsonSerializerOptions options)
    {
        writer.WriteStringValue(value.Item.ToString());
    }
}

public class SentimentTypeConverter : GenericConverterWithToString<SentimentType>
{
    public override SentimentType Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        return SentimentType.FromString(reader.GetString());
    }
}

public class TrendDirectionConverter : GenericConverterWithToString<Trends.TrendDirection>
{
    public override Trends.TrendDirection Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        return Trends.TrendDirection.FromString(reader.GetString());
    }
}

public class TrendTypeConverter : GenericConverterWithToString<Trends.TrendType>
{
    public override Trends.TrendType Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        return Trends.TrendType.FromString(reader.GetString());
    }
}

public class OrderStatusConverter : GenericConverterWithToString<OrderStatus>
{
    public override OrderStatus Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        throw new NotImplementedException("OrderStatus is not deserializable");
    }
}

public class OptionOrderTypeConverter : GenericConverterWithToString<OptionOrderType>
{
    public override OptionOrderType Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        throw new NotImplementedException("OrderType is not deserializable");
    }
}

public class StockOrderTypeConverter : GenericConverterWithToString<StockOrderType>
{
    public override StockOrderType Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        throw new NotImplementedException("OrderType is not deserializable");
    }
}

public class StockOrderInstructionConverter : GenericConverterWithToString<StockOrderInstruction>
{
    public override StockOrderInstruction Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        throw new NotImplementedException("OrderInstruction is not deserializable");
    }
}

public class OptionOrderInstructionConverter : GenericConverterWithToString<OptionOrderInstruction>
{
    public override OptionOrderInstruction Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        throw new NotImplementedException("OrderInstruction is not deserializable");
    }
}

public class AssetTypeConverter : GenericConverterWithToString<AssetType>
{
    public override AssetType Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        throw new NotImplementedException("AssetType is not deserializable");
    }
}

public class AccountTransactionTypeConverter : GenericConverterWithToString<AccountTransactionType>
{
    public override AccountTransactionType Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        throw new NotImplementedException();
    }
}

public class StockTransactionTypeConverter : GenericConverterWithToString<StockTransactionType>
{
    public override StockTransactionType Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        throw new NotImplementedException();
    }
}

public class OptionTypeTypeConverter : JsonConverter<OptionType>
{
public override OptionType Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
{
    return OptionType.FromString(reader.GetString());
}
    
public override void Write(Utf8JsonWriter writer, OptionType value, JsonSerializerOptions options)
{
    writer.WriteStringValue(value.ToString());
}
}

public class OptionPositionIdConverter : JsonConverter<OptionPositionId>
{
    public override OptionPositionId Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        return OptionPositionId.NewOptionPositionId(Guid.Parse(reader.GetString()));
    }
    
    public override void Write(Utf8JsonWriter writer, OptionPositionId value, JsonSerializerOptions options)
    {
        writer.WriteStringValue(value.Item.ToString());
    }
}

public class OptionExpirationConverter : JsonConverter<OptionExpiration>
{
    public override OptionExpiration Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        return OptionExpiration.create(reader.GetString());
    }
    
    public override void Write(Utf8JsonWriter writer, OptionExpiration value, JsonSerializerOptions options)
    {
        writer.WriteStringValue(value.ToString());
    }
}

File: src/web.fs/Utils/CustomConverters.fs
namespace web.Utils

open System
open System.Text.Json
open System.Text.Json.Serialization
open core.fs
open core.fs.Accounts
open core.fs.Adapters.Brokerage
open core.fs.Adapters.Stocks
open core.fs.Alerts
open core.fs.Options
open core.fs.Reports
open core.fs.Services
open core.fs.Services.Analysis
open core.fs.Stocks
open core.Shared
open core.Stocks

[<AbstractClass>]
type GenericConverterWithToString<'T>() =
    inherit JsonConverter<'T>()
    
    override this.Write(writer: Utf8JsonWriter, value: 'T, options: JsonSerializerOptions) =
        writer.WriteStringValue(value.ToString())

type TickerConverter() =
    inherit JsonConverter<Ticker>()
    
    override this.Read(reader: byref<Utf8JsonReader>, typeToConvert: Type, options: JsonSerializerOptions) =
        Ticker(reader.GetString())
    
    override this.Write(writer: Utf8JsonWriter, value: Ticker, options: JsonSerializerOptions) =
        writer.WriteStringValue(value.Value)

type OptionTickerConverter() =
    inherit JsonConverter<OptionTicker>()
    
    override this.Read(reader: byref<Utf8JsonReader>, typeToConvert: Type, options: JsonSerializerOptions) =
        OptionTicker(reader.GetString())
    
    override this.Write(writer: Utf8JsonWriter, value: OptionTicker, options: JsonSerializerOptions) =
        let (OptionTicker str) = value
        writer.WriteStringValue(str)

type UserIdConverter() =
    inherit JsonConverter<UserId>()
    
    override this.Read(reader: byref<Utf8JsonReader>, typeToConvert: Type, options: JsonSerializerOptions) =
        UserId(Guid(reader.GetString()))
    
    override this.Write(writer: Utf8JsonWriter, value: UserId, options: JsonSerializerOptions) =
        let (UserId guid) = value
        writer.WriteStringValue(guid)

type GapTypeConverter() =
    inherit GenericConverterWithToString<GapAnalysis.GapType>()
    
    override this.Read(reader: byref<Utf8JsonReader>, typeToConvert: Type, options: JsonSerializerOptions) =
        GapAnalysis.GapType.FromString(reader.GetString())

type PriceFrequencyConverter() =
    inherit GenericConverterWithToString<PriceFrequency>()
    
    override this.Read(reader: byref<Utf8JsonReader>, typeToConvert: Type, options: JsonSerializerOptions) =
        PriceFrequency.FromString(reader.GetString())

type OutcomeReportDurationConverter() =
    inherit GenericConverterWithToString<OutcomesReportDuration>()
    
    override this.Read(reader: byref<Utf8JsonReader>, typeToConvert: Type, options: JsonSerializerOptions) =
        OutcomesReportDuration.FromString(reader.GetString())

type OptionTypeConverter() =
    inherit GenericConverterWithToString<OptionType>()
    
    override this.Read(reader: byref<Utf8JsonReader>, typeToConvert: Type, options: JsonSerializerOptions) =
        OptionType.FromString(reader.GetString())

type GradeConverter() =
    inherit JsonConverter<TradeGrade>()
    
    override this.Read(reader: byref<Utf8JsonReader>, typeToConvert: Type, options: JsonSerializerOptions) =
        TradeGrade(reader.GetString())
    
    override this.Write(writer: Utf8JsonWriter, value: TradeGrade, options: JsonSerializerOptions) =
        writer.WriteStringValue(value.Value)

type ValueFormatTypeConverter() =
    inherit GenericConverterWithToString<ValueFormat>()
    
    override this.Read(reader: byref<Utf8JsonReader>, typeToConvert: Type, options: JsonSerializerOptions) =
        ValueFormat.FromString(reader.GetString())

type OutcomeTypeTypeConverter() =
    inherit GenericConverterWithToString<OutcomeType>()
    
    override this.Read(reader: byref<Utf8JsonReader>, typeToConvert: Type, options: JsonSerializerOptions) =
        OutcomeType.FromString(reader.GetString())

type BrokerageOrderTypeConverter() =
    inherit GenericConverterWithToString<BrokerageOrderType>()
    
    override this.Read(reader: byref<Utf8JsonReader>, typeToConvert: Type, options: JsonSerializerOptions) =
        BrokerageOrderType.FromString(reader.GetString())

type BrokerageOrderDurationConverter() =
    inherit GenericConverterWithToString<BrokerageOrderDuration>()
    
    override this.Read(reader: byref<Utf8JsonReader>, typeToConvert: Type, options: JsonSerializerOptions) =
        BrokerageOrderDuration.FromString(reader.GetString())

type PositionEventTypeConverter() =
    inherit JsonConverter<PositionEventType>()
    
    override this.Read(reader: byref<Utf8JsonReader>, typeToConvert: Type, options: JsonSerializerOptions) =
        PositionEventType(reader.GetString())
    
    override this.Write(writer: Utf8JsonWriter, value: PositionEventType, options: JsonSerializerOptions) =
        writer.WriteStringValue(value.Value)

type ChartAnnotationLineTypeConverter() =
    inherit GenericConverterWithToString<ChartAnnotationLineType>()
    
    override this.Read(reader: byref<Utf8JsonReader>, typeToConvert: Type, options: JsonSerializerOptions) =
        ChartAnnotationLineType.FromString(reader.GetString())

type DataPointChartTypeConverter() =
    inherit GenericConverterWithToString<DataPointChartType>()
    
    override this.Read(reader: byref<Utf8JsonReader>, typeToConvert: Type, options: JsonSerializerOptions) =
        DataPointChartType.FromString(reader.GetString())

type StockPositionIdConverter() =
    inherit JsonConverter<StockPositionId>()
    
    override this.Read(reader: byref<Utf8JsonReader>, typeToConvert: Type, options: JsonSerializerOptions) =
        StockPositionId(Guid.Parse(reader.GetString()))
    
    override this.Write(writer: Utf8JsonWriter, value: StockPositionId, options: JsonSerializerOptions) =
        let (StockPositionId guid) = value
        writer.WriteStringValue(guid.ToString())

type SentimentTypeConverter() =
    inherit GenericConverterWithToString<SentimentType>()
    
    override this.Read(reader: byref<Utf8JsonReader>, typeToConvert: Type, options: JsonSerializerOptions) =
        SentimentType.FromString(reader.GetString())

type TrendDirectionConverter() =
    inherit GenericConverterWithToString<InflectionPoints.TrendDirection>()
    
    override this.Read(reader: byref<Utf8JsonReader>, typeToConvert: Type, options: JsonSerializerOptions) =
        InflectionPoints.TrendDirection.FromString(reader.GetString())

type InflectionPointsTypeConverter() =
    inherit GenericConverterWithToString<InflectionPoints.InfectionPointType>()
    
    override this.Read(reader: byref<Utf8JsonReader>, typeToConvert: Type, options: JsonSerializerOptions) =
        InflectionPoints.InfectionPointType.FromString(reader.GetString())

type TrendTypeConverter() =
    inherit GenericConverterWithToString<Trends.TrendType>()
    
    override this.Read(reader: byref<Utf8JsonReader>, typeToConvert: Type, options: JsonSerializerOptions) =
        Trends.TrendType.FromString(reader.GetString())

type OrderStatusConverter() =
    inherit GenericConverterWithToString<OrderStatus>()
    
    override this.Read(reader: byref<Utf8JsonReader>, typeToConvert: Type, options: JsonSerializerOptions) =
        raise (NotImplementedException("OrderStatus is not deserializable"))

type OptionOrderTypeConverter() =
    inherit GenericConverterWithToString<OptionOrderType>()
    
    override this.Read(reader: byref<Utf8JsonReader>, typeToConvert: Type, options: JsonSerializerOptions) =
        raise (NotImplementedException("OrderType is not deserializable"))

type StockOrderTypeConverter() =
    inherit GenericConverterWithToString<StockOrderType>()
    
    override this.Read(reader: byref<Utf8JsonReader>, typeToConvert: Type, options: JsonSerializerOptions) =
        raise (NotImplementedException("OrderType is not deserializable"))

type StockOrderInstructionConverter() =
    inherit GenericConverterWithToString<StockOrderInstruction>()
    
    override this.Read(reader: byref<Utf8JsonReader>, typeToConvert: Type, options: JsonSerializerOptions) =
        raise (NotImplementedException("OrderInstruction is not deserializable"))

type OptionOrderInstructionConverter() =
    inherit GenericConverterWithToString<OptionOrderInstruction>()
    
    override this.Read(reader: byref<Utf8JsonReader>, typeToConvert: Type, options: JsonSerializerOptions) =
        raise (NotImplementedException("OrderInstruction is not deserializable"))

type AssetTypeConverter() =
    inherit GenericConverterWithToString<AssetType>()
    
    override this.Read(reader: byref<Utf8JsonReader>, typeToConvert: Type, options: JsonSerializerOptions) =
        raise (NotImplementedException("AssetType is not deserializable"))

type AccountTransactionTypeConverter() =
    inherit GenericConverterWithToString<AccountTransactionType>()
    
    override this.Read(reader: byref<Utf8JsonReader>, typeToConvert: Type, options: JsonSerializerOptions) =
        raise (NotImplementedException())

type StockTransactionTypeConverter() =
    inherit GenericConverterWithToString<StockTransactionType>()
    
    override this.Read(reader: byref<Utf8JsonReader>, typeToConvert: Type, options: JsonSerializerOptions) =
        raise (NotImplementedException())

type OptionTypeTypeConverter() =
    inherit JsonConverter<OptionType>()
    
    override this.Read(reader: byref<Utf8JsonReader>, typeToConvert: Type, options: JsonSerializerOptions) =
        OptionType.FromString(reader.GetString())
    
    override this.Write(writer: Utf8JsonWriter, value: OptionType, options: JsonSerializerOptions) =
        writer.WriteStringValue(value.ToString())

type OptionPositionIdConverter() =
    inherit JsonConverter<OptionPositionId>()
    
    override this.Read(reader: byref<Utf8JsonReader>, typeToConvert: Type, options: JsonSerializerOptions) =
        OptionPositionId(Guid.Parse(reader.GetString()))
    
    override this.Write(writer: Utf8JsonWriter, value: OptionPositionId, options: JsonSerializerOptions) =
        let (OptionPositionId guid) = value
        writer.WriteStringValue(guid.ToString())

type OptionExpirationConverter() =
    inherit JsonConverter<OptionExpiration>()
    
    override this.Read(reader: byref<Utf8JsonReader>, typeToConvert: Type, options: JsonSerializerOptions) =
        OptionExpiration.create(reader.GetString())
    
    override this.Write(writer: Utf8JsonWriter, value: OptionExpiration, options: JsonSerializerOptions) =
        writer.WriteStringValue(value.ToString())
